src/orion/algo/tpe.py

```diff
@@ -0,0 +1,396 @@
+        from the import space and return them.
+
+        :param num: how many sets to be suggested.
+
+        .. note:: New parameters must be compliant with the problem's domain
+           `orion.algo.space.Space`.
+        """
+        if num > 1:
+            raise ValueError("TPE should suggest only one point.")
+
+        samples = []
+        if len(self._trials_info) < self.n_initial_points:
+            new_point = self.space.sample(1, seed=tuple(self.rng.randint(0, 1000000, size=3)))[0]
+            samples.append(new_point)
+        else:
+            point = []
+            below_points, above_points = self.split_trials()
+            below_points = numpy.array([flatten_dims(point, self.space) for point in below_points])
+            above_points = numpy.array([flatten_dims(point, self.space) for point in above_points])
+
+            idx = 0
+            for dimension in self.space.values():
+
+                shape = dimension.shape
+                if not shape:
+                    shape = (1,)
+
+                if dimension.type == 'real':
+                    points = self.sample_real_dimension(dimension, shape[0],
+                                                        below_points[:, idx: idx + shape[0]],
+                                                        above_points[:, idx: idx + shape[0]])
+                else:
+                    raise ValueError("TPE now only support Real Dimension.")
+
+                if len(points) < shape[0]:
+                    logger.warning('TPE failed to sample new point with configuration %s',
+                                   self.configuration)
+                    return None
+
+                idx += shape[0]
+                point += points
+
+            point = regroup_dims(point, self.space)
+            samples.append(point)
+
+        return samples
+
+    def sample_real_dimension(self, dimension, shape_size, below_points, above_points):
+        """Sample values for a real dimension
+
+        :param dimension: Real Dimension.
+        :param shape_size: 1D Shape Size of the Real Dimension.
+        :param below_points: good points with shape (m, n), n=shape_size.
+        :param above_points: bad points with shape (m, n), n=shape_size.
+        """
+        points = []
+
+        for j in range(shape_size):
+            new_point = self._sample_real_point(dimension, below_points[:, j], above_points[:, j])
+            if new_point:
+                points.append(new_point)
+
+        return points
+
+    def _sample_real_point(self, dimension, below_points, above_points):
+        """Sample one value for a real dimension based on the observed good and bad points"""
+        low, high = dimension.interval()
+        below_mus, below_sigmas, below_weights = \
+            adaptive_parzen_estimator(below_points, low, high, self.prior_weight,
+                                      self.equal_weight, flat_num=25)
```

Comment:
Shouldn't flat_num be a hyper-parameter of TPE? Maybe we could merge equal_weight with flat_num; if flat_num is None or equal 0, we use equal weights, otherwise we use the ramp up with flat_num.

Reply:
I was thinking to keep this parameter as internal at this moment, unlike equal_weight, this parameter has more impact on the performance, want to hide it with best practice value mentioned in the paper.

Reply:
The value of 25 may very well be domain specific however. They did test on different ones (LFW, PubFig83 and subset of CIFAR10), but that's arguably small evidence.

Reply:
okay, let's then expose an additional parameter for tpe.